### DISCUSSION TITLE
Upcoming removal of org.graalvm.nativeimage:svm from maven?

### DISCUSSION THREAD

**Discussion by zakkak** [oracle/graal#4146] [General]
In https://graalvm.slack.com/archives/CNBFR78F9/p1639154644225700 @gilles-duboscq referring to `org.graalvm.nativeimage:svm` mentioned that:

> hum... we thought that was already deprecated in 21,3 and were planning to remove it in 22.0. But it was indeed still deployed with 21.3

As discussed in that slack thread there are projects relying on some of the APIs provided by it.

For instance, Quarkus is currently using the following imports from that jar and its removal would be a breaking change:
```
import com.oracle.svm.core.annotate.Alias;
import com.oracle.svm.core.annotate.AlwaysInline;
import com.oracle.svm.core.annotate.AutomaticFeature;
import com.oracle.svm.core.annotate.Delete;
import com.oracle.svm.core.annotate.Inject;
import com.oracle.svm.core.annotate.InjectAccessors;
import com.oracle.svm.core.annotate.KeepOriginal;
import com.oracle.svm.core.annotate.NeverInline;
import com.oracle.svm.core.annotate.RecomputeFieldValue;
import com.oracle.svm.core.annotate.RecomputeFieldValue.Kind;
import com.oracle.svm.core.annotate.Substitute;
import com.oracle.svm.core.annotate.TargetClass;
import com.oracle.svm.core.annotate.TargetElement;
import com.oracle.svm.core.jdk.JDK11OrLater;
import com.oracle.svm.core.jdk.JDK17OrLater;
import com.oracle.svm.core.jdk.JDK8OrEarlier;
import com.oracle.svm.core.posix.headers.Pthread;
import com.oracle.svm.core.SubstrateUtil;
import com.oracle.svm.core.threadlocal.FastThreadLocalFactory;
import com.oracle.svm.core.threadlocal.FastThreadLocalInt;
```

I am starting this discussion aiming to find a way to allow the usage of such APIs even after the removal of the `svm` artifact (possibly by moving them in `sdk`).

**Comment by kirillp** [comment]
We have huge list of portability examples. Java is designed to be portable. The core question here is that GraalVM compiler introduced a new better way to JNI. This is cool, however, HotSpot has other focus-pocus (see Java critical), TeaVM-C has its own annotations related to native interop. In order to get all things together svm annotations are essential.

**Comment by kirillp** [comment]
Another point is that GraalVM is very slow, it is [12x times slower](https://github.com/oracle/graal/issues/2873) doing jni call with arrays then jvm, and use of SVM annotations makes this a little bit better

**Comment by gilles-duboscq** [comment]
For 22.0 specifically, we can release the SVM jar as we did for 21.3: it's built with Java 8 and might not work very well on Java 11+.
I think in the cases you mention above you only use this jar to compile against some internal classes and you don't really need the code inside that jar to really work to build native-images. So i think those use-cases will be fine.

It remains a question of what to do moving forward.

Note that the SVM and compiler JARs are definitely not stable APIs. They only contain internal classes that might change in arbitrary way between each release.
They were provided on maven to help with very specific edge use-cases (the old native-image plugin, using the compiler in a non-graalvm JDK to run graal-js).
The Graal SDK was created to provide stable APIs that users can rely on.

**Reply by zakkak** [reply]
> For 22.0 specifically, we can release the SVM jar as we did for 21.3: it's built with Java 8 and might not work very well on Java 11+.

I think that would be the safest thing to do for now. After all it has already been "tested" with 21.3 :)

> I think in the cases you mention above you only use this jar to compile against some internal classes and you don't really need the code inside that jar to really work to build native-images. So i think those use-cases will be fine.

Correct, we only need access to the API so that we can compile Quarkus with any JDK and not rely on GraalVM for that.

> It remains a question of what to do moving forward.
>
> Note that the SVM and compiler JARs are definitely not stable APIs. They only contain internal classes that might change in arbitrary way between each release.

IMO we should find a way to make the functionality provided by the annotations available through some public API. I understand that exposing these (especially Substitutions) might be a bit dangerous, but these annotations are the ones enabling Quarkus to work around issues with libraries that are not native-ready and to perform optimizations by reducing the reachable classes/objects in certain cases, etc.

**Reply by gilles-duboscq** [reply]
> these annotations are the ones enabling Quarkus to work around issues with libraries that are not native-ready and to perform optimizations by reducing the reachable classes/objects in certain cases, etc.

Working around issues in libraries that are not native-ready and whose source code cannot be modified is indeed the use-case for substitutions.

Regarding reducing reachability for optimizations, do you also mean in such libraries that are outside your control? Or do you have cases where you need substitutions even though you would be able to modify the code of the classes targeted by substitutions?

**Reply by kirillp** [reply]
Yes we have such cases.
When a team works on project it uses jvm mostly. And so we use native jni methods to reach c code. With GraalVM it is 12-50x times slower then using annotations. So we use substitute to substitute native mathods with CFunctions or to avoid jni with pinned objects or ObjectHandles. Which are unavaible for other Java platforms:  HotSpot, Android, TeaVM-C, TeaVM-JS

**Reply by zakkak** [reply]
> Regarding reducing reachability for optimizations, do you also mean in such libraries that are outside your control?

Correct.

> Or do you have cases where you need substitutions even though you would be able to modify the code of the classes targeted by substitutions?

No, in cases where we can alter the code, or ask the upstream project to do so, we don't use substitutions (unless for quick-fixes till upstream resolves the issue).

**Reply by kirillp** [reply]
@zakkak you can avoid using annotations because you have the only target, if you had really multiplatfom code that targets 5-6 platforms you'll have to use em

**Reply by zakkak** [reply]
> @zakkak you can avoid using annotations because you have the only target, if you had really multiplatfom code that targets 5-6 platforms you'll have to use em

That's true we only support OpenJDK-HotSpot and GraalVM/Mandrel.

What you describe sounds like something that could also be handled by providing two implementations of the API (hopefully only the native part), one suitable for GraalVM and another suitable for the rest of the platforms. What are the benefits of using substitutions instead of providing two implementations?

**Reply by kirillp** [reply]
I dont see how this is **effectively** possible
Assume you have 
```
public class Dispatch {
  static native void dispatchAsync(long threadIsolate, Runnable r);
}

```
which is located in module platform and used in 7 other modules
and there is class DispatchReplacement that usses grall specific things to implement it

```
@TargetClass(value = Dispatch.class)
public final class DispatchReplacement {
  public static void dispatchAsync(long isolate, Runnable r) {
    ObjectHandle objectHandle = handles.create(r);
    dispatch_async(
        isolate,
        objectHandle.rawValue(),
        Mth.callBack.getFunctionPointer().rawValue());
  }
}

```
or another example: 

```
 public final class GL {
  public static native void glFinish();
 }
```

which is substitued by 

```
  @CFunction(transition=NO_TRANSITION)
  public static native void glFinish();

```

And if you have this

```
public class ObjC {
 public static native long sel_getUid(byte[] name);
}
```

the problem is described here 
https://github.com/oracle/graal/issues/2873
the only way to solve is to use 

```
@TargetClass(value = ObjC.class)
public final class ObjCReplacement {
  @Substitute
  public static long sel_getUid(byte[] name) {
    PinnedObject p = PinnedObject.create(name);
    long uid = ObjCNative.sel_getUid(p.addressOfArrayElement(0));
    p.close();
    return uid;
  }
}

```

OF course we can avoid using the Annotations by introducing two new modules for each substitutions and substitute one module by another ie: compile module A with generic code, while linking (or packaging) the app with A_Graal_Specific instead. But this is pain

Instead one can substitute and feel good.

**Reply by gilles-duboscq** [reply]
So the second issue here is that you need to use `PinnedObject` to work around JNI performance issues on SVM.
I don't really understand the first one though, how do you implement that on HotSpot? Can't you use the same JNI code on SVM?

**Reply by kirillp** [reply]
Jni emulation is extremely slow in GraalVM. JNI C iterface emulation is not that good. 
Using it in practice makes performance problems. Anotations does the job. Same time other vms does not have such annotations. So we substitute JNI with SVM

**Comment by kirillp** [comment]
A solution is to move annotations to a separate module.

**Comment by kirillp** [comment]
> I think in the cases you mention above you only use this jar to compile against some internal classes and you don't really need the code inside that jar to really work to build native-images

Exactly, we need only annotations, not the code.

**Comment by kirillp** [comment]
About the following inports  I agree with @gilles-duboscq point here
`
import com.oracle.svm.core.jdk.JDK11OrLater;
import com.oracle.svm.core.jdk.JDK17OrLater;
import com.oracle.svm.core.jdk.JDK8OrEarlier;
import com.oracle.svm.core.posix.headers.Pthread;
import com.oracle.svm.core.SubstrateUtil;
import com.oracle.svm.core.threadlocal.FastThreadLocalFactory;
import com.oracle.svm.core.threadlocal.FastThreadLocalInt;
`

One should not use class like com.oracle.svm.core.* 
It is internal implementation details.
If one needs the API he can declare it same way.

**Comment by kristofdho** [comment]
A few classes we're using for configuration inside custom `Feature` implementations:
`import com.oracle.svm.core.jni.JNIRuntimeAccess;`: JNI configuration during analysis should be configurable through the SDK. It is available for reflection, with the `RuntimeReflection` class.
`import com.oracle.svm.core.configure.ResourcesRegistry;`: Same argument as `JNIRuntimeAccess`.
`import com.oracle.svm.hosted.FeatureImpl;`: to access calls like `FeatureImpl.DuringSetupAccessImpl#registerClassReachabilityListener` to act on reachable classes with a specific annotation. But could be used to act on any arbitrary class matching.

**Comment by jerboaa** [comment]
So I've done a comparison of a JDK 11 build `svm.jar` and a JDK 17 build `svm.jar`. Note that `svm.jar` is the artefact discussed for removal in this discussion. Note: I've actually done a diff of all GraalVM specific jars I could find (of a mandrel distro) and only `svm.jar` shows differences. As expected there are certain classes not present in an earlier JDK-based build that would be available in a later JDK-based build. That is because the build system builds certain classes only conditionally (based on the base JDK version). If people start using the artefact with different JDK versions things are bound to break. A solution for that problem is indeed needed.

```
--- jdk11_list.txt	2022-01-20 20:45:37.258706026 +0100
+++ jdk17_list.txt	2022-01-20 20:45:46.846747729 +0100
@@ -2230,6 +2230,24 @@
 ./svm.jar.d/com/oracle/svm/core/jdk11/Target_jdk_internal_loader_BuiltinClassLoader.class
 ./svm.jar.d/com/oracle/svm/core/jdk11/Target_jdk_internal_loader_ClassLoaders_JDK11OrLater.class
 ./svm.jar.d/com/oracle/svm/core/jdk11/Target_sun_misc_Unsafe.class
+./svm.jar.d/com/oracle/svm/core/jdk14
+./svm.jar.d/com/oracle/svm/core/jdk14/Target_java_nio_DirectByteBuffer_JDK14OrLater.class
+./svm.jar.d/com/oracle/svm/core/jdk14/Target_java_nio_DirectByteBufferR_JDK14OrLater.class
+./svm.jar.d/com/oracle/svm/core/jdk14/Target_sun_nio_ch_Util.class
+./svm.jar.d/com/oracle/svm/core/jdk15
+./svm.jar.d/com/oracle/svm/core/jdk15/HiddenClassFeature.class
+./svm.jar.d/com/oracle/svm/core/jdk15/HiddenClassSupportImpl.class
+./svm.jar.d/com/oracle/svm/core/jdk15/Target_java_lang_ClassLoader_JDK15OrLater.class
+./svm.jar.d/com/oracle/svm/core/jdk15/Target_java_lang_Module_JDK15OrLater.class
+./svm.jar.d/com/oracle/svm/core/jdk15/Target_jdk_internal_loader_BootLoader_JDK15OrLater.class
+./svm.jar.d/com/oracle/svm/core/jdk15/Target_jdk_internal_loader_NativeLibraries_JDK15OrLater.class
+./svm.jar.d/com/oracle/svm/core/jdk15/Target_jdk_internal_loader_NativeLibraries_NativeLibraryImpl_JDK15OrLater.class
+./svm.jar.d/com/oracle/svm/core/jdk16
+./svm.jar.d/com/oracle/svm/core/jdk16/JDKClassLoadingDeadlockWorkaround.class
+./svm.jar.d/com/oracle/svm/core/jdk16/RecordFeatureJDK16OrLater.class
+./svm.jar.d/com/oracle/svm/core/jdk16/RecordSupportJDK16OrLater.class
+./svm.jar.d/com/oracle/svm/core/jdk17
+./svm.jar.d/com/oracle/svm/core/jdk17/Target_jdk_internal_platform_CgroupMetrics_JDK17OrLater.class
 ./svm.jar.d/com/oracle/svm/core/jdk/AccessControlContextFeature.class
 ./svm.jar.d/com/oracle/svm/core/jdk/AccessControllerUtil.class
 ./svm.jar.d/com/oracle/svm/core/jdk/AllPermissionsPolicy.class
@@ -4196,6 +4214,7 @@
 ./svm.jar.d/com/oracle/svm/hosted/jdk/JNIRegistrationsJavaZip.class
 ./svm.jar.d/com/oracle/svm/hosted/jdk/JNIRegistrationSupport$1.class
 ./svm.jar.d/com/oracle/svm/hosted/jdk/JNIRegistrationSupport.class
+./svm.jar.d/com/oracle/svm/hosted/jdk/JNIRegistrationSupport_JDK14OrLater.class
 ./svm.jar.d/com/oracle/svm/hosted/jdk/package-info.class
 ./svm.jar.d/com/oracle/svm/hosted/lambda
 ./svm.jar.d/com/oracle/svm/hosted/lambda/LambdaProxyRenamingSubstitutionProcessor.class
```

**Reply by kirillp** [reply]
As I already mentioned above the solution is trivial - make a new module `annotations` or move annotations to the sdk so that people can remove the svm dependency

**Reply by gilles-duboscq** [reply]
A solution we have used for this issue previously is multi-release jars.

In this case it will only make a difference if JDK and `svm.jar` version are mixed when actually building native-images. If `svm.jar` is only used to compile against some internal APIs i don't think it's an issue since none of the classes shown here are likely to be used in this way.

**Reply by jerboaa** [reply]
IMHO, multi-release jars seem a better approach than the status quo for `svm.jar` ;-) I agree it's unlikely to be hit in practice, but still something worth cleaning up (i.e. make it harder for people to make the mistake to begin with).

**Reply by gilles-duboscq** [reply]
It is [now](https://github.com/oracle/graal/commit/413017d336aa0fefa39a87d405b72a5ffd399fe0#diff-3b099932fe18682ea0ff68228170442fc0a1fe8398ded31e54ce210f1c4ab546R56) a 11+17 multi-release jar.

**Comment by kristofdho** [comment]
It's unclear to me if the svm jar is still going to be released or not. If not, this should get some more attention as according to https://github.com/oracle/graal/discussions/4310 the feature freeze for the next release will be on March 11 already.

**Reply by gilles-duboscq** [reply]
As far as I can see, in the current state we don't have much choice and we will keep pushing the SVM jar out. It is now a 11+17 multi-release jar.

**Reply by meltsufin** [reply]
Is version 22.1.0 going to be released?

**Reply by gilles-duboscq** [reply]
Yes, #4518 is tracking the missing SVM jars on maven for 22.1.0

**Comment by gilles-duboscq** [comment]
For reference, another usage of `@AutomaticFeature` can be seen in #4518
